src/lshashtable.c

```diff
@@ -0,0 +1,826 @@
+}
+
+static inline struct lsht_record_store_rec *lsht_record_store_clone(struct lshashtable *lsht, struct lsht_record_store *rs, struct lsht_record_store_rec *rec) {
+    size_t size = _rec_size(rec);
+    lshashtable_make_room(lsht, rs, size + rs->min_free_size);
+    if (lsht_record_store_avail(rs) < size)
+        return NULL;
+    struct lsht_record_store_rec *newrec = rs_wloc(rs);
+    _lsht_record_store_set_use(rs, _lsht_record_store_pos2blockpos(rs_wlocpos(rs)));
+    memcpy(newrec, rec, size);
+    rs_wseek(rs, size);
+    rs->num_free_blocks -= newrec->num_blocks;
+    return newrec;
+}
+
+static inline uint16_t lshashtable_get_rs_id(struct lshashtable *lsht, struct lsht_record_store *rs) {
+    struct lsht_record_store *rs_begin = (struct lsht_record_store *)vmbuf_data(&lsht->record_stores);
+    return rs - rs_begin;
+}
+
+static inline void *_next_rec(void *r) {
+    return r + (((struct lsht_record_store_rec *)r)->num_blocks << RS_BLOCK_SIZE_BITS);
+}
+
+static inline int lshashtable_is_expired(struct lshashtable *lsht, struct lshashtable_rec *rec) {
+    return rec->val_size != UINT_MAX && lsht->is_expired(rec);
+}
+
+static inline void lshashtable_make_room(struct lshashtable *lsht, struct lsht_record_store *rs, size_t size) {
+    uint64_t num_deleted_keys = 0;
+    struct vmbuf *recs = &rs->recbuf;
+    vmbuf_init(recs, 4096);
+    uint16_t rs_id = lshashtable_get_rs_id(lsht, rs), next_rs_id = rs_id + 1;
+    size_t num_free_blocks_threshold = (size >> RS_BLOCK_SIZE_BITS) + (lsht_record_store_total_blocks(rs) >> 2);
+    while (lsht_record_store_avail(rs) < size) {
+        /* reclaim space of deleted records */
+        while (!rs_empty(rs)) {
+            struct lsht_record_store_rec *r = lsht_record_store_oldest_rec(rs);
+            if (_lsht_record_store_is_inuse(rs, _lsht_record_store_pos2blockpos(rs_rlocpos(rs)))) {
+                struct lshashtable_rec *rec = (struct lshashtable_rec *)r->data;
+                /* if the record is the special record for deleted key and in the last store, we can safely discard it */
+                if (UINT_MAX != rec->val_size || next_rs_id != lsht->next_rs_id) {
+                    if (lshashtable_is_expired(lsht, rec)) {
+                        lshashtable_del_key_index(lsht, rec->data, rec->key_size, NULL);
+                        lsht_record_store_del(rs, rec);
+                    } else
+                        break;
+                }
+            }
+            rs_rseek(rs, _rec_size(r));
+        }
+        if (lsht_record_store_avail(rs) >= size)
+            break;
+        /* sanity check */
+        if (rs_empty(rs)) {
+            LOGGER_ERROR("requested size is larger than configured capacity: %zu > %zu, aborting...", size, rs_capacity(rs));
+            abort();
+        }
+        struct lsht_record_store_rec *r = lsht_record_store_pop_oldest_rec(rs);
+        if (lsht_record_store_num_free_blocks(rs) > num_free_blocks_threshold) { /* relocate */
```

Comment:
Good example. It means that k2 record is not frequently updated relatively to others and its expiration time is too long for this store full cycle so we better keep it out of the way in the next level store otherwise we will stumble over it all the time. On practice it is likely that we would have more records like that from expiration time and access pattern perspective so the store will not be wasted. I think it would be more consistent approach and will keep LRU order, simplify the code and need less computing cycles work (record moves and updates to the index that more frequent in stores with smaller index).  It is probably ok to keep it like that for disk base store since but for memory based case I think I will add code to avoid relocation not to relocate. For this PR  we can keep it as it is for both cases.